addons/test_mail/tests/test_mail_followers_unsubscribe.py

```diff
@@ -48,3 +54,156 @@ def test_inbox_notification_follower(self):
+            return self._mails[0]['body']
+
+    def _notify_thread_by_email(self, record, partner):
+        with self.mock_mail_gateway():
+            recipients_data = self._generate_notify_recipients(partner)
+            test_message = self.env['mail.message'].browse(self.test_message.ids).copy()
+            record._notify_thread_by_email(test_message, recipients_data)
+            self.assertEqual(len(self._mails), 1)
+            return self._mails[0]['body']
+
+    def _url_tampering(self, url, **kwargs):
+        parsed_url = urlparse(url)
+        return parsed_url._replace(query=urlencode(dict(parse_qsl(parsed_url.query), **kwargs))).geturl()
+
+    def _test_tampered_unfollow_url(self, record, unfollow_url, partner):
+        for param, value in (('token', '0000000000000000000000000000000000000000'),
+                             ('model', 'mail.test.gateway'),
+                             ('res_id', record.copy().id)):
+            with self.subTest(f'Tampered {param}'):
+                tampered_unfollow_url = self._url_tampering(unfollow_url, **{param: value})
+                response = self.url_open(tampered_unfollow_url)
+                self.assertEqual(response.status_code, 403)
+                self.assertIn(partner, record.message_partner_ids)
+
+        with self.subTest('Tampered partner id'):
+            record._message_subscribe(partner_ids=self.partner_admin.ids)
+            tampered_unfollow_url = self._url_tampering(unfollow_url, pid=self.partner_admin.id)
+            response = self.url_open(tampered_unfollow_url)
+            self.assertEqual(response.status_code, 403)
+            self.assertIn(partner, record.message_partner_ids)
+            self.assertIn(self.partner_admin, record.message_partner_ids)
+            record.message_unsubscribe(partner_ids=self.partner_admin.ids)
+
+    def _test_unfollow_url(self, record, unfollow_url, partner, landing_url_check_mode):
+        with self.subTest('Legitimate unfollow'):
+            # We test that the URL still work a second time if the user has been re-added
+            for _ in range(2):
+                try:
+                    self.assertIn(partner, record.message_partner_ids)
+                    response = self.url_open(unfollow_url)
+                    self.assertEqual(response.status_code, 200)
+                    self.assertNotIn(partner, record.message_partner_ids.ids)
+                    if landing_url_check_mode == 'record':
+                        self.assertEqual(urlparse(response.url).path, '/web')
+                        self.assertIn(f'model={record._name}', response.url)
+                        self.assertIn(f'id={record.id}', response.url)
+                    elif landing_url_check_mode == 'front_end_message':
+                        self.assertEqual(urlparse(response.url).path, '/mail/unfollow')
+                        self.assertIn("You are no longer following the document.", response.text)
+                    else:
```

Comment:
can we instead add at the start
```py
self.assertIn(landing_url_check_mode, ('record', 'front_end_message'))
```

Reply:
Yes good idea. Added at the start and I have removed the else clause with the exception.